File: src/scpn_fusion/control/spi_ablation.py
# ──────────────────────────────────────────────────────────────────────
# SCPN Fusion Core — SPI Ablation Solver
# © 1998–2026 Miroslav Šotek. All rights reserved.
# ──────────────────────────────────────────────────────────────────────
"""
Multi-fragment Shattered Pellet Injection (SPI) Ablation Solver.

Tracks the trajectories and ablation of neutral gas fragments injected
into the plasma. Based on the Parks ablation model.

Physics:
- Lagrangian tracking of N fragments (position r, velocity v).
- Shielding-modifed Parks ablation rate:
  dm/dt = -1.25e16 * n_e^(1/3) * T_e^(1.64) * r_p^(1.33) / v_p
  (Reduced-order scaling calibrated for fast control studies)
- Deposition of ablated material into the plasma density profile.
"""

from __future__ import annotations

import logging
import numpy as np
from dataclasses import dataclass
from typing import List, Tuple

logger = logging.getLogger(__name__)

# Physical constants
AMU = 1.66e-27             # atomic mass unit [kg]
M_NEON = 20.18 * AMU       # neon atom mass [kg]
RHO_NEON_SOLID = 1444.0    # solid neon density [kg/m^3], CRC Handbook

# Parks ablation rate prefactor [g/s] in code units (ne in 10^20 m^-3, Te in keV, rp in cm).
# Derived from Parks NF 57 (2017) Eq. 8: C_P = 1.25e16 (CGS), converted to mixed units
# via dimensional analysis: C_P_mixed = C_P * (1e20)^{-1/3} * keV^{-1.64} * cm^{-1.33} ≈ 2.0.
_PARKS_COEFFICIENT = 2.0

# Tokamak geometry (ITER-like)
R_MAJOR = 6.2              # major radius [m]
A_MINOR = 2.0              # minor radius [m]
ELONGATION = 2.0           # plasma elongation

@dataclass
class SpiFragment:
    """A single shard of the shattered pellet."""
    id: int
    radius: float # m
    mass: float   # kg
    pos: np.ndarray # (x, y, z) [m]
    vel: np.ndarray # (vx, vy, vz) [m/s]
    active: bool = True

class SpiAblationSolver:
    def __init__(self, 
                 n_fragments: int = 100, 
                 total_mass_kg: float = 0.01, # 10g Neon
                 velocity_mps: float = 200.0,
                 dispersion: float = 0.1, # Velocity spread fraction
                 injector_pos: np.ndarray | None = None, # Outboard midplane
                 injector_dir: np.ndarray | None = None, # Radial inward
                 seed: int = 42,
                 ):
        if isinstance(n_fragments, bool) or int(n_fragments) < 1:
            raise ValueError("n_fragments must be an integer >= 1.")
        total_mass_kg = float(total_mass_kg)
        velocity_mps = float(velocity_mps)
        dispersion = float(dispersion)
        if not np.isfinite(total_mass_kg) or total_mass_kg <= 0.0:
            raise ValueError("total_mass_kg must be finite and > 0.")
        if not np.isfinite(velocity_mps) or velocity_mps <= 0.0:
            raise ValueError("velocity_mps must be finite and > 0.")
        if not np.isfinite(dispersion) or dispersion < 0.0:
            raise ValueError("dispersion must be finite and >= 0.")
        if isinstance(seed, bool) or not isinstance(seed, (int, np.integer)):
            raise ValueError("seed must be an integer.")

        n_fragments = int(n_fragments)
        injector_pos_arr = np.asarray(
            [10.0, 0.0, 0.0] if injector_pos is None else injector_pos,
            dtype=np.float64,
        )
        injector_dir_arr = np.asarray(
            [-1.0, 0.0, 0.0] if injector_dir is None else injector_dir,
            dtype=np.float64,
        )
        if injector_pos_arr.shape != (3,) or not np.all(np.isfinite(injector_pos_arr)):
            raise ValueError("injector_pos must be a finite 3-vector.")
        if injector_dir_arr.shape != (3,) or not np.all(np.isfinite(injector_dir_arr)):
            raise ValueError("injector_dir must be a finite 3-vector.")
        dir_norm = float(np.linalg.norm(injector_dir_arr))
        if dir_norm <= 0.0:
            raise ValueError("injector_dir must be non-zero.")

        self.fragments: List[SpiFragment] = []

        # Uniform fragment mass; size from solid-sphere volume
        m_frag = total_mass_kg / n_fragments
        vol = m_frag / RHO_NEON_SOLID
        r_frag = (3 * vol / (4 * np.pi))**(1/3)

        rng = np.random.default_rng(int(seed))
        for i in range(n_fragments):
            v_dir = injector_dir_arr + rng.normal(0, dispersion, 3)
            v_dir /= np.linalg.norm(v_dir)
            v_mag = velocity_mps * rng.normal(1.0, 0.1)
            vel = v_dir * v_mag
            pos = injector_pos_arr + rng.normal(0, 0.05, 3)
            
            self.fragments.append(SpiFragment(
                id=i,
                radius=r_frag,
                mass=m_frag,
                pos=pos,
                vel=vel
            ))
            
    def step(self, dt: float, plasma_ne_profile: np.ndarray, plasma_te_profile: np.ndarray, r_grid: np.ndarray) -> np.ndarray:
        """Advance fragments and return deposition profile [particles/m^3/s]."""
        dt = float(dt)
        if not np.isfinite(dt) or dt <= 0.0:
            raise ValueError("dt must be finite and > 0.")
        plasma_ne_profile = np.asarray(plasma_ne_profile, dtype=np.float64)
        plasma_te_profile = np.asarray(plasma_te_profile, dtype=np.float64)
        r_grid = np.asarray(r_grid, dtype=np.float64)
        if plasma_ne_profile.ndim != 1 or plasma_te_profile.ndim != 1 or r_grid.ndim != 1:
            raise ValueError("plasma_ne_profile, plasma_te_profile, and r_grid must be 1D arrays.")
        if len(r_grid) < 2:
            raise ValueError("r_grid must have at least 2 points.")
        if len(plasma_ne_profile) != len(plasma_te_profile):
            raise ValueError("plasma_ne_profile and plasma_te_profile must have matching lengths.")
        if not np.all(np.isfinite(plasma_ne_profile)) or not np.all(np.isfinite(plasma_te_profile)):
            raise ValueError("plasma profiles must be finite.")
        if not np.all(np.isfinite(r_grid)) or not np.all(np.diff(r_grid) > 0.0):
            raise ValueError("r_grid must be finite and strictly increasing.")

        deposition = np.zeros_like(r_grid)
        dr = r_grid[1] - r_grid[0] if len(r_grid) > 1 else 1.0

        for frag in self.fragments:
            if not frag.active:
                continue

            frag.pos += frag.vel * dt

            r_loc = np.sqrt(frag.pos[0]**2 + frag.pos[1]**2)
            z_loc = frag.pos[2]
            # Map (R,Z) to normalised minor radius rho
            rho_loc = np.sqrt(
                ((r_loc - R_MAJOR) / A_MINOR)**2 + (z_loc / ELONGATION)**2
            )

            if rho_loc > 1.2 or rho_loc < 0.0:
                continue

            rho_axis = np.linspace(0, 1, len(plasma_ne_profile))
            n_e = np.interp(rho_loc, rho_axis, plasma_ne_profile)   # [10^19 m^-3]
            T_e = np.interp(rho_loc, np.linspace(0, 1, len(plasma_te_profile)), plasma_te_profile)  # [keV]

            if T_e < 0.01:
                continue

            # Parks ablation scaling, Parks NF 57 (2017) Eq. 8
            ne_20 = max(float(n_e) / 10.0, 0.0)
            rp_cm = frag.radius * 100.0
            dm_dt_g = _PARKS_COEFFICIENT * (ne_20**0.33) * (float(T_e)**1.64) * (rp_cm**1.33)
            dm_dt_kg = dm_dt_g / 1000.0

            delta_m = dm_dt_kg * dt
            if delta_m > frag.mass:
                delta_m = frag.mass
                frag.active = False
                frag.mass = 0.0
                frag.radius = 0.0
            else:
                frag.mass -= delta_m
                frag.radius = (3 * (frag.mass / RHO_NEON_SOLID) / (4 * np.pi))**(1/3)

            n_particles = delta_m / M_NEON
            idx = int(round(rho_loc * (len(r_grid) - 1)))
            if 0 <= idx < len(r_grid):
                rate = n_particles / dt
                r_minor = rho_loc * A_MINOR
                dV = 4 * np.pi**2 * R_MAJOR * r_minor * (A_MINOR * dr)
                if dV < 1e-3:
                    dV = 1.0  # on-axis guard
                deposition[idx] += rate / dV

        return deposition

File: scpn-fusion-rs/crates/fusion-control/src/spi_ablation.rs
// ─────────────────────────────────────────────────────────────────────
// SCPN Fusion Core — SPI Fragment Ablation Solver
// © 1998–2026 Miroslav Šotek. All rights reserved.
// ─────────────────────────────────────────────────────────────────────
//! Multi-fragment Shattered Pellet Injection (SPI) ablation solver.
//!
//! Port of `spi_ablation.py`.
//! Lagrangian tracking of N fragments with Parks ablation model.
//! Struct-of-arrays layout for cache efficiency.
//!
//! Reference: Parks, NF 57 (2017) Eq. 8.

use fusion_types::error::{FusionError, FusionResult};
use rand::prelude::*;
use rand_distr::Normal;
use std::f64::consts::PI;

/// Neon atom mass [kg], 20.18 AMU.
const M_NEON: f64 = 20.18 * 1.66e-27;
/// Solid neon density [kg/m^3], CRC Handbook.
const RHO_NEON_SOLID: f64 = 1444.0;
/// Parks ablation prefactor, Parks NF 57 (2017) Eq. 8 in mixed code units.
const PARKS_COEFFICIENT: f64 = 2.0;
/// Major radius [m], ITER-like.
const R_MAJOR: f64 = 6.2;
/// Minor radius [m].
const A_MINOR: f64 = 2.0;
/// Plasma elongation.
const ELONGATION: f64 = 2.0;

/// SPI fragment ablation solver — struct-of-arrays layout.
pub struct SpiAblationSolver {
    pub n_fragments: usize,
    pub radius: Vec<f64>,
    pub mass: Vec<f64>,
    pub pos_x: Vec<f64>,
    pub pos_y: Vec<f64>,
    pub pos_z: Vec<f64>,
    pub vel_x: Vec<f64>,
    pub vel_y: Vec<f64>,
    pub vel_z: Vec<f64>,
    pub active: Vec<bool>,
}

impl SpiAblationSolver {
    pub fn new(
        n_fragments: usize,
        total_mass_kg: f64,
        velocity_mps: f64,
        dispersion: f64,
    ) -> FusionResult<Self> {
        if n_fragments == 0 {
            return Err(FusionError::ConfigError(
                "n_fragments must be > 0".to_string(),
            ));
        }
        if !total_mass_kg.is_finite() || total_mass_kg <= 0.0 {
            return Err(FusionError::ConfigError(
                "total_mass must be finite and > 0".to_string(),
            ));
        }

        let m_frag = total_mass_kg / n_fragments as f64;
        let vol = m_frag / RHO_NEON_SOLID;
        let r_frag = (3.0 * vol / (4.0 * PI)).cbrt();

        let mut rng = StdRng::seed_from_u64(42);
        let dir_noise = Normal::new(0.0, 1.0)
            .map_err(|e| FusionError::ConfigError(e.to_string()))?;
        let vel_noise = Normal::new(1.0, 0.1)
            .map_err(|e| FusionError::ConfigError(e.to_string()))?;
        let pos_noise = Normal::new(0.0, 0.05)
            .map_err(|e| FusionError::ConfigError(e.to_string()))?;

        let mut solver = SpiAblationSolver {
            n_fragments,
            radius: vec![r_frag; n_fragments],
            mass: vec![m_frag; n_fragments],
            pos_x: vec![0.0; n_fragments],
            pos_y: vec![0.0; n_fragments],
            pos_z: vec![0.0; n_fragments],
            vel_x: vec![0.0; n_fragments],
            vel_y: vec![0.0; n_fragments],
            vel_z: vec![0.0; n_fragments],
            active: vec![true; n_fragments],
        };

        // Outboard midplane injector, radial inward
        for i in 0..n_fragments {
            let mut vd = [
                -1.0 + dir_noise.sample(&mut rng) * dispersion,
                dir_noise.sample(&mut rng) * dispersion,
                dir_noise.sample(&mut rng) * dispersion,
            ];
            let norm = (vd[0] * vd[0] + vd[1] * vd[1] + vd[2] * vd[2])
                .sqrt()
                .max(1e-30);
            vd[0] /= norm;
            vd[1] /= norm;
            vd[2] /= norm;

            let v_mag = velocity_mps * vel_noise.sample(&mut rng);
            solver.vel_x[i] = vd[0] * v_mag;
            solver.vel_y[i] = vd[1] * v_mag;
            solver.vel_z[i] = vd[2] * v_mag;

            solver.pos_x[i] = 10.0 + pos_noise.sample(&mut rng);
            solver.pos_y[i] = pos_noise.sample(&mut rng);
            solver.pos_z[i] = pos_noise.sample(&mut rng);
        }

        Ok(solver)
    }

    /// Advance fragments by dt. Returns deposition profile [particles/m^3/s].
    pub fn step(
        &mut self,
        dt: f64,
        plasma_ne: &[f64],
        plasma_te: &[f64],
        r_grid: &[f64],
    ) -> FusionResult<Vec<f64>> {
        let n_grid = r_grid.len();
        let mut deposition = vec![0.0; n_grid];
        let dr = if n_grid > 1 {
            r_grid[1] - r_grid[0]
        } else {
            1.0
        };

        let rho_ne = linspace(0.0, 1.0, plasma_ne.len());
        let rho_te = linspace(0.0, 1.0, plasma_te.len());

        for i in 0..self.n_fragments {
            if !self.active[i] {
                continue;
            }

            self.pos_x[i] += self.vel_x[i] * dt;
            self.pos_y[i] += self.vel_y[i] * dt;
            self.pos_z[i] += self.vel_z[i] * dt;

            let r_loc = (self.pos_x[i].powi(2) + self.pos_y[i].powi(2)).sqrt();
            let z_loc = self.pos_z[i];
            let rho_loc = (((r_loc - R_MAJOR) / A_MINOR).powi(2)
                + (z_loc / ELONGATION).powi(2))
            .sqrt();

            if !(0.0..=1.2).contains(&rho_loc) {
                continue;
            }

            let n_e = interp(rho_loc, &rho_ne, plasma_ne);
            let t_e = interp(rho_loc, &rho_te, plasma_te);

            if t_e < 0.01 {
                continue;
            }

            // Parks ablation, Parks NF 57 (2017) Eq. 8
            let ne_20 = (n_e / 10.0).max(0.0);
            let rp_cm = self.radius[i] * 100.0;
            let dm_dt_g =
                PARKS_COEFFICIENT * ne_20.powf(0.33) * t_e.powf(1.64) * rp_cm.powf(1.33);
            let dm_dt_kg = dm_dt_g / 1000.0;

            let mut delta_m = dm_dt_kg * dt;
            if delta_m > self.mass[i] {
                delta_m = self.mass[i];
                self.active[i] = false;
                self.mass[i] = 0.0;
                self.radius[i] = 0.0;
            } else {
                self.mass[i] -= delta_m;
                self.radius[i] = (3.0 * (self.mass[i] / RHO_NEON_SOLID) / (4.0 * PI)).cbrt();
            }

            let n_particles = delta_m / M_NEON;
            let idx = (rho_loc * (n_grid - 1) as f64).round() as usize;
            if idx < n_grid {
                let rate = n_particles / dt;
                let r_minor = rho_loc * A_MINOR;
                let dv = 4.0 * PI * PI * R_MAJOR * r_minor * (A_MINOR * dr);
                deposition[idx] += rate / dv.max(1.0);
            }
        }

        Ok(deposition)
    }

    pub fn n_active(&self) -> usize {
        self.active.iter().filter(|&&a| a).count()
    }

    pub fn total_mass(&self) -> f64 {
        self.mass.iter().sum()
    }
}

fn linspace(start: f64, end: f64, n: usize) -> Vec<f64> {
    if n <= 1 {
        return vec![start];
    }
    (0..n)
        .map(|i| start + (end - start) * i as f64 / (n - 1) as f64)
        .collect()
}

fn interp(x: f64, xp: &[f64], fp: &[f64]) -> f64 {
    if xp.is_empty() {
        return 0.0;
    }
    if x <= xp[0] {
        return fp[0];
    }
    if x >= *xp.last().unwrap() {
        return *fp.last().unwrap();
    }
    let mut lo = 0;
    let mut hi = xp.len() - 1;
    while hi - lo > 1 {
        let mid = (lo + hi) / 2;
        if xp[mid] <= x {
            lo = mid;
        } else {
            hi = mid;
        }
    }
    let t = (x - xp[lo]) / (xp[hi] - xp[lo]).max(1e-30);
    fp[lo] + t * (fp[hi] - fp[lo])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_creation() {
        let s = SpiAblationSolver::new(100, 0.01, 200.0, 0.1).unwrap();
        assert_eq!(s.n_fragments, 100);
        assert_eq!(s.n_active(), 100);
        assert!((s.total_mass() - 0.01).abs() < 1e-12);
    }

    #[test]
    fn test_step_returns_deposition() {
        let mut s = SpiAblationSolver::new(100, 0.01, 200.0, 0.1).unwrap();
        let ne = vec![5.0; 50];
        let te = vec![10.0; 50];
        let r_grid: Vec<f64> = (0..50).map(|i| i as f64 / 49.0).collect();
        let dep = s.step(1e-4, &ne, &te, &r_grid).unwrap();
        assert_eq!(dep.len(), 50);
    }

    #[test]
    fn test_mass_decreases() {
        let mut s = SpiAblationSolver::new(10, 0.01, 200.0, 0.1).unwrap();
        let ne = vec![5.0; 50];
        let te = vec![10.0; 50];
        let r_grid: Vec<f64> = (0..50).map(|i| i as f64 / 49.0).collect();
        let m0 = s.total_mass();
        for _ in 0..100 {
            s.step(1e-4, &ne, &te, &r_grid).unwrap();
        }
        assert!(s.total_mass() <= m0);
    }

    #[test]
    fn test_rejects_invalid() {
        assert!(SpiAblationSolver::new(0, 0.01, 200.0, 0.1).is_err());
        assert!(SpiAblationSolver::new(100, -1.0, 200.0, 0.1).is_err());
    }
}
